pkg/apiserver/cubeapi/user/auth.go

```diff
@@ -156,3 +119,123 @@ func Login(c *gin.Context) {
 	response.SuccessReturn(c, user)
 	return
 }
+
+func GitHubLogin(c *gin.Context) {
+	code := c.Query("code")
+	if code == "" {
+		clog.Error("code is null")
+		response.FailReturn(c, errcode.AuthenticateError)
+		return
+	}
+
+	g := github.GetProvider()
+	if !g.GitHubIsEnable {
+		clog.Error("github auth is disabled")
+		response.FailReturn(c, errcode.AuthenticateError)
+		return
+	}
+
+	userInfo, err := g.IdentityExchange(code)
+	if err != nil {
+		response.FailReturn(c, errcode.AuthenticateError)
+		return
+	}
+	clog.Info("user %s auth success by github", userInfo.GetUserName())
+
+	// get user by name
+	userName := hex.EncodeToString([]byte(userInfo.GetUserName()))
+	userFind, respInfo := GetUserByName(c, gitHubUserNamePrefix+userName)
+	if respInfo != nil {
+		response.FailReturn(c, respInfo)
+		return
+	}
+	if userFind != nil && userFind.Spec.State == v1.ForbiddenState {
+		response.FailReturn(c, errcode.UserIsDisabled)
+		return
+	}
+
+	// if user first login, create user
+	user := &v1.User{}
+	if userFind == nil {
+		user.Name = gitHubUserNamePrefix + userName
+		user.Spec.LoginType = v1.GitHubLogin
+		user.Spec.Password = md5util.GetMD5Salt(uuid.New().String())
```

Comment:
We are supposed to add rand seed if we want to use random(rand.Seed(time.Now().UnixNano()))